baseplate/lib/message_queue.py

```diff
@@ -111,14 +150,119 @@ def unlink(self) -> None:
-        Python.
 
-        """
-        self.queue.close()
+class InMemoryMessageQueue(MessageQueue):
+    """An in-memory inter process message queue.
+
+    Uses a simple Python Queue to store data.
+
+    """
+
+    def __init__(self, max_messages: int):
+        self.queue: q.Queue = q.Queue(max_messages)
+        self.max_messages = max_messages
+
+    def get(self, timeout: Optional[float] = None) -> bytes:
+        try:
+            message = self.queue.get(timeout=timeout)
+            self.queue.task_done()
+            return message
+        except q.Empty:
+            raise TimedOutError
+
+    def put(self, message: bytes, timeout: Optional[float] = None) -> None:
+        try:
+            self.queue.put(message, timeout=timeout)
+        except q.Full:
+            raise TimedOutError
+
+    def close(self) -> None:
+        """Not implemented for in-memory queue"""
+
+
+class RemoteMessageQueue(MessageQueue):
+    """A message queue that uses a remote Thrift server.
+
+    Used in conjunction with the InMemoryMessageQueue to
+    provide an alternative to Posix queues, for use-cases
+    where Posix is not appropriate.
+
+    """
+
+    def __init__(
+        self,
+        name: str,
+        max_messages: int,
+        host: str = DEFAULT_QUEUE_HOST,
+        port: int = DEFAULT_QUEUE_PORT,
+    ):
+        # Connect to the remote queue server, and creeate the new queue
+        self.name = name
+        self.max_messages = max_messages
+        self.host = host
+        self.port = port
+        self.connect()
+        self.client.create_queue(name, max_messages)
+
+    def connect(self) -> None:
+        # Establish a connection with the queue server
+        transport = TSocket.TSocket(self.host, self.port)
+        self.transport = TTransport.TBufferedTransport(transport)
+        protocol = TBinaryProtocol.TBinaryProtocol(self.transport)
+        self.client = RemoteMessageQueueService.Client(protocol)
+        self.transport.open()
+
+    def _try_to_get(self, timeout: Optional[float]) -> bytes:
+        try:
+            return self.client.get(self.name, timeout).value
+        except TSocket.TTransportException:
+            self.connect()  # Try reconnecting once
+            return self.client.get(self.name, timeout).value
+
+    def get(self, timeout: Optional[float] = None) -> bytes:
+        # Call the remote server and get an element for the correct queue
+        try:
+            return self._try_to_get(timeout)
+        # If the server responded with a timeout, raise our own timeout to be consistent with the posix queue
+        except ThriftTimedOutError:
+            raise TimedOutError
+
+    def _try_to_put(self, message: bytes, timeout: Optional[float]) -> None:
+        try:
+            self.client.put(self.name, message, timeout)
+        except TSocket.TTransportException:
+            self.connect()  # Try reconnecting once
+            self.client.put(self.name, message, timeout)
```

Comment:
Looking at this for the first time today, so I may be missing something, but if I'm reading this correctly, on the client side, we're coupling putting a message on the queue with a TCP network request to send a Thrift (or possibly HTTP in the future) message to the sidecar. Unless I"m missing something, this will force the client to wait for an acknowledgment from the server, both at TCP level as well as the l7 protocol level. This is a significant change in behavior from how the POSIX queue is going to behave, where the messages are, outside of exceptional cases, going to be always acknowledged by the queue -- now we're dependent on server behavior. If the server degrades in any way, the service emitting events is going to see latency impact, or worse.

Running against a local sidecar will hide latency issues in the common case, but if the sidecar server wedges, or anything like that, we're going to be in trouble.

If we need delivery guarantees here that are leading us to use TCP and protocols that provide message acknowledgement, I think we'll need to do some client side buffering of these events for this so we can decouple the main baseplate app from the behavior of the server. This is going to increase the memory footprint of every baseplate application though because we have to account for that size of that message buffer now, but that should be an acceptable tradeoff.

Again, this is just based on my initial reading here, so I might be wrong.